include/psa/crypto_se_driver.h

```diff
@@ -961,6 +961,83 @@ typedef struct {
 
 /**@}*/
 
+/** \defgroup se_registration Secure element driver registration
+ */
+/**@{*/
+
+/** A structure containing pointers to all the entry points of a
+ * secure element driver.
+ *
+ * Future versions of this specification may add extra substructures at
+ * the end of this structure.
+ */
+typedef struct {
+    /** The version of the driver model that this driver implements.
+     * This is a protection against linking driver binaries built against
+     * a different version of this specification.
+     * Use #PSA_DRV_SE_HAL_VERSION.
```

Comment:
Could you explain a bit more what you are protecting against? Perhaps you are thinking that there is a potential misuse when parts of the code are taken from different versions of the library and linked together? Is this a concern at all or something that would happen often? At the moment it looks like a bit of an overkill to me...

Reply:
We need version so we can support multiple versions of drivers if we break the API or change behavior. A driver can use the version field to request a specific set of behaviors from the core library. A library may say "sorry, that version isn't supported or compatible", or it might provide backwards compatible behaviors. If we don't have the version field from the beginning, we paint ourselves into a corner with fewer options for the evolution of the driver model.

Reply:
Ok, I think I understand the need for a mechanism like this, but does it need to be a runtime check? If we only expect PSA to be used in small embedded systems where there is no dynamic linking, then we could probably just do it statically assuming that the driver exports some sort of version macro (not an unreasonable assumption in my opinion). But I do not think this works in all cases if we expect PSA and drivers to be loaded dynamically.

@gilles-peskine-arm made a comment about this [here](https://github.com/ARMmbed/mbed-crypto/pull/155#discussion_r297579898).